[KV] Drafting a more detailed Overview page for Workers KV. (#19830)
cloudflare/cloudflare-docs: src/content/docs/kv/index.mdx

<Plan type="workers-all" />

> ## What is Workers KV?
> 

Reviewer: ```suggestion
```

			//delete a key-value pair
			await env.KV_BINDING.delete('KEY');

			//return a Workers response
			return new Response(
				JSON.stringify({
					value: value,
					allKeys: allKeys,
				}),
			);
		},
	};

	```
	</TabItem>
	<TabItem label="wrangler.jsonc">
		```json
		{
			"$schema": "node_modules/wrangler/config-schema.json",
			"name": "WORKER-NAME",
			"main": "src/index.ts",
			"compatibility_date": "2025-02-04",
			"observability": {
				"enabled": true
			},

			"kv_namespaces": [
				{
					"binding": "YOUR_BINDING",
					"id": "<YOUR_BINDING_ID>"
				}
			]
		```
	</TabItem>
	</Tabs>
</TabItem>
<TabItem label="REST API">
```
curl https://api.cloudflare.com/client/v4/accounts/$ACCOUNT_ID/storage/kv/namespaces/$NAMESPACE_ID/values/$KEY_NAME \
    -X PUT \
    -H 'Content-Type: multipart/form-data' \
    -H "X-Auth-Email: $CLOUDFLARE_EMAIL" \
    -H "X-Auth-Key: $CLOUDFLARE_API_KEY" \
    -d '{
      "value": "Some Value"
    }'

curl https://api.cloudflare.com/client/v4/accounts/$ACCOUNT_ID/storage/kv/namespaces/$NAMESPACE_ID/values/$KEY_NAME \
    -H "X-Auth-Email: $CLOUDFLARE_EMAIL" \
    -H "X-Auth-Key: $CLOUDFLARE_API_KEY"
```
</TabItem>
</Tabs>

> Learn more about Workers KV [key concepts](/kv/concepts/how-kv-works/), or [get started](/kv/get-started/) with a Workers project.

Reviewer: ```suggestion
Learn more about Workers KV [key concepts](/kv/concepts/how-kv-works/), or [get started](/kv/get-started/) with a Cloudflare Workers project.
```

<Plan type="workers-all" />

## What is Workers KV?

Workers KV is a data storage that allows you to store and retrieve data globally. With Workers KV, you can build dynamic and performant APIs and websites that support high read volumes with low latency.

> Workers KV is ideal for projects that require:
> 
> - High volumes of reads and/or repeated reads to the same keys.
> - Per-object time-to-live (TTL).
> - Distributed configuration.
> 
> For example, you can use Workers KV for:
> 
> - Caching API responses.
> - Storing user configurations / preferences.
> - Storing user authentication details.

Reviewer: ```suggestion

For example, you can use Workers KV for:

- Caching API responses.
- Storing user configurations / preferences.
- Storing user authentication details.

Workers KV is ideal for projects that require:

- High volumes of reads and/or repeated reads to the same keys.
- Per-object time-to-live (TTL).
- Distributed configuration.
```

<Tabs>
<TabItem label="index.ts">
```ts
export interface Env {
	YOUR_BINDING: KVNamespace;
}

> export default {
> 	async fetch(request, env, ctx): Promise<Response> {
> 		await env.YOUR_BINDING.put('KEY', 'VALUE');
> 		const value = await env.YOUR_BINDING.get('KEY');
> 		const allKeys = await env.YOUR_BINDING.list();
> 		await env.YOUR_BINDING.delete('KEY');
> 		return new Response(
> 			JSON.stringify({
> 				value: value,
> 				allKeys: allKeys,
> 			}),
> 			{
> 				headers: {
> 					'Content-Type': 'application/json'
> 				},
> 			}
> 		);
> 	},
> };

Reviewer: ```suggestion
export default {
	async fetch(request, env, ctx): Promise<Response> {
	
	  //write a key-value pair
		await env.YOUR_BINDING.put('KEY', 'VALUE');
		
		//read a key-value pair
		const value = await env.YOUR_BINDING.get('KEY');
		
		//list all key-value pairs
		const allKeys = await env.YOUR_BINDING.list();
		
		//delete a key-value pair
		await env.YOUR_BINDING.delete('KEY');
		
		
		//return a Workers response
		return new Response(
			JSON.stringify({
				value: value,
				allKeys: allKeys,
			})
			}
		);
	},
};
```

<Plan type="workers-all" />

## What is Workers KV?

Workers KV is a data storage that allows you to store and retrieve data globally. With Workers KV, you can build dynamic and performant APIs and websites that support high read volumes with low latency.

For example, you can use Workers KV for:

- Caching API responses.
- Storing user configurations / preferences.
- Storing user authentication details.

Workers KV is ideal for projects that require:

- High volumes of reads and/or repeated reads to the same keys.
- Globally distributed data.

> **Code examples**

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
Access your Workers KV namespace from Cloudflare Workers using the Workers Binding API or from your external application using the REST API:
```

Author: Actioned outside GH

<Tabs>
<TabItem label="REST API">
```
curl https://api.cloudflare.com/client/v4/accounts/$ACCOUNT_ID/storage/kv/namespaces/$NAMESPACE_ID/values/$KEY_NAME \
    -X PUT \
    -H 'Content-Type: multipart/form-data' \
    -H "X-Auth-Email: $CLOUDFLARE_EMAIL" \
    -H "X-Auth-Key: $CLOUDFLARE_API_KEY" \
    -d '{
      "value": "Some Value"
    }'

curl https://api.cloudflare.com/client/v4/accounts/$ACCOUNT_ID/storage/kv/namespaces/$NAMESPACE_ID/values/$KEY_NAME \
    -H "X-Auth-Email: $CLOUDFLARE_EMAIL" \
    -H "X-Auth-Key: $CLOUDFLARE_API_KEY"
```
</TabItem>
<TabItem label="Workers Binding API">
```ts

> export interface Env {
> 	YOUR_BINDING: KVNamespace;
> }

Reviewer: ```suggestion
```

Author: Actioned outside GH

> export default {
> 	async fetch(request, env, ctx): Promise<Response> {
> 
> 		//write a key-value pair
> 		await env.YOUR_BINDING.put('KEY', 'VALUE');
> 
> 		// read a key-value pair
> 		const value = await env.YOUR_BINDING.get('KEY');
> 
> 		//list all key-value pairs
> 		const allKeys = await env.YOUR_BINDING.list();
> 
> 		//delete a key-value pair
> 		await env.YOUR_BINDING.delete('KEY');
> 
> 		//return a Workers response
> 		return new Response(
> 			JSON.stringify({
> 				value: value,
> 				allKeys: allKeys,
> 			}),
> 		);
> 	},
> };

Reviewer: ```suggestion
export default {
	async fetch(request, env, ctx): Promise<Response> {

		//write a key-value pair
		await env.KV_BINDING.put('KEY', 'VALUE');

		// read a key-value pair
		const value = await env.KV_BINDING.get('KEY');

		//list all key-value pairs
		const allKeys = await env.KV_BINDING.list();

		//delete a key-value pair
		await env.KV_BINDING.delete('KEY');

		//return a Workers response
		return new Response(
			JSON.stringify({
				value: value,
				allKeys: allKeys,
			}),
		);
	},
};
```

> <TabItem label="REST API">
> ```
> curl https://api.cloudflare.com/client/v4/accounts/$ACCOUNT_ID/storage/kv/namespaces/$NAMESPACE_ID/values/$KEY_NAME \
>     -X PUT \
>     -H 'Content-Type: multipart/form-data' \
>     -H "X-Auth-Email: $CLOUDFLARE_EMAIL" \
>     -H "X-Auth-Key: $CLOUDFLARE_API_KEY" \
>     -d '{
>       "value": "Some Value"
>     }'
> 
> curl https://api.cloudflare.com/client/v4/accounts/$ACCOUNT_ID/storage/kv/namespaces/$NAMESPACE_ID/values/$KEY_NAME \
>     -H "X-Auth-Email: $CLOUDFLARE_EMAIL" \
>     -H "X-Auth-Key: $CLOUDFLARE_API_KEY"
> ```
> </TabItem>
> <TabItem label="Workers Binding API">
> ```ts
> export interface Env {
> 	YOUR_BINDING: KVNamespace;
> }
> 
> export default {
> 	async fetch(request, env, ctx): Promise<Response> {
> 
> 		//write a key-value pair
> 		await env.YOUR_BINDING.put('KEY', 'VALUE');
> 
> 		// read a key-value pair
> 		const value = await env.YOUR_BINDING.get('KEY');
> 
> 		//list all key-value pairs
> 		const allKeys = await env.YOUR_BINDING.list();
> 
> 		//delete a key-value pair
> 		await env.YOUR_BINDING.delete('KEY');
> 
> 		//return a Workers response
> 		return new Response(
> 			JSON.stringify({
> 				value: value,
> 				allKeys: allKeys,
> 			}),
> 		);
> 	},
> };
> 
> ```
> </TabItem>

Reviewer: ```suggestion
<TabItem label="Workers Binding API">
```ts
export interface Env {
	YOUR_BINDING: KVNamespace;
}

export default {
	async fetch(request, env, ctx): Promise<Response> {

		//write a key-value pair
		await env.YOUR_BINDING.put('KEY', 'VALUE');

		// read a key-value pair
		const value = await env.YOUR_BINDING.get('KEY');

		//list all key-value pairs
		const allKeys = await env.YOUR_BINDING.list();

		//delete a key-value pair
		await env.YOUR_BINDING.delete('KEY');

		//return a Workers response
		return new Response(
			JSON.stringify({
				value: value,
				allKeys: allKeys,
			}),
		);
	},
};

```
</TabItem>
<TabItem label="REST API">
```
curl https://api.cloudflare.com/client/v4/accounts/$ACCOUNT_ID/storage/kv/namespaces/$NAMESPACE_ID/values/$KEY_NAME \
    -X PUT \
    -H 'Content-Type: multipart/form-data' \
    -H "X-Auth-Email: $CLOUDFLARE_EMAIL" \
    -H "X-Auth-Key: $CLOUDFLARE_API_KEY" \
    -d '{
      "value": "Some Value"
    }'

curl https://api.cloudflare.com/client/v4/accounts/$ACCOUNT_ID/storage/kv/namespaces/$NAMESPACE_ID/values/$KEY_NAME \
    -H "X-Auth-Email: $CLOUDFLARE_EMAIL" \
    -H "X-Auth-Key: $CLOUDFLARE_API_KEY"
```
</TabItem>

```

> <TabItem label="wrangler.json">
> ```json
> {
> 	"$schema": "node_modules/wrangler/config-schema.json",
> 	"name": "WORKER-NAME",
> 	"main": "src/index.ts",
> 	"compatibility_date": "2025-02-04",
> 	"observability": {
> 		"enabled": true
> 	},
> 
> 	"kv_namespaces": [
> 		{
> 			"binding": "YOUR_BINDING",
> 			"id": "<YOUR_BINDING_ID>"
> 		}
> 	]
> ```
> </TabItem>

Reviewer: ```suggestion
```

<Plan type="workers-all" />

## What is Workers KV?

Workers KV is a data storage that allows you to store and retrieve data globally. With Workers KV, you can build dynamic and performant APIs and websites that support high read volumes with low latency.

For example, you can use Workers KV for:

- Caching API responses.
- Storing user configurations / preferences.
- Storing user authentication details.

Workers KV is ideal for projects that require:

> - High volumes of reads and/or repeated reads to the same keys.
> - Per-object time-to-live (TTL).
> - Distributed configuration.

Reviewer: ```suggestion
- High volumes of reads and/or repeated reads to the same keys.
- Geo-distributed data with built-in caching (per-object time-to-live (TTL)).
```
